### PR TITLE
feat: Implement RDF/JS dataset specification

### LINKED ISSUE #390: Support for set operations on graphs/datasets in the N3Store
Hi, I was wondering if it's within the scope of N3.js to support set operations on graphs or datasets within the N3Store?

If I am able to contribute an implementation, would you accept it? Or would an extension like this fall outside the scope of N3.js?

### LINKED ISSUE #391: Why does `#every` return false on the empty store
As shown in the test here `#every` returns `false` when called on an empty store. There is a test enforcing this behavior at https://github.com/rdfjs/N3.js/blob/0ba9ec2e0e7d812f9e86985de3966583d228e189/test/N3Store-test.js#L92 This does not align with set semantics where `#every` would align with `forAll` and thus be vacuously `true` on the empty set.

It also goes against the RDF/JS Dataset definition of for `#every` https://rdf.js.org/dataset-spec/#dom-dataset-every; and so in order to complete https://github.com/rdfjs/N3.js/pull/288 we would need to alter this behavior and decide whether this change is (1) breaking or (2) a bug fix - despite the fact that there is a test enforcing the buggy behavior. @RubenVerborgh what do you think?

### PR DESCRIPTION
**Pull request by jeswr**
Draft PR to add full functionality of https://github.com/rdfjs/types/blob/master/dataset.d.ts. Observe that we alter the behavior of `#empty` on the empty set to now return `true` instead of `false`, as described in https://github.com/rdfjs/N3.js/issues/391 - so it may be that this needs to be an mver bump.

Supercedes https://github.com/rdfjs/N3.js/pull/288, pushing to a branch on the N3.js repo so CI will run.

Closes #390
Closes #391

### CODE REVIEW

File: src/N3Store.js [hunk 1]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+      this.removeQuad(quad);
+    return this;
+  }
+
+  /**
+   * Returns a new dataset that contains all quads from the current dataset, not included in the given dataset.
+   */
+  difference(other) {
+    return this.filter(quad => !other.has(quad));
+  }
```
Comment (jeswr) [inline src/N3Store.js:823]: There is lots of room for optimisation here; in particular, whenever `other` is also an `N3Store` and they share an index we can just merge / diff / compare the indices for set operations.

Would be worth opening up an issue to do this once this PR is merged.

File: README.md [hunk 2]
```diff
@@ -318,7 +318,7 @@ Methods:
  - `[Symbol.iterator]` — Implements the iterator protocol to allow iteration over all `quads` in the dataset as in the example above.
 
 ### Addition and deletion of quads
-The store provides the following manipulation methods in addition to implementing the standard [`DatasetCore` Interface](https://rdf.js.org/dataset-spec/#datasetcore-interface)
+The store provides the following manipulation methods in addition to implementing the standard [`Dataset` Interface](https://rdf.js.org/dataset-spec/#dataset-interface)
```
Comment (TallTed) [inline README.md:321]:
```suggestion
The store implements the following manipulation methods in addition to the standard [`Dataset` Interface](https://rdf.js.org/dataset-spec/#dataset-interface)
```

File: src/N3Store.js [hunk 3]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
     return lists;
   }
 
+  /**
+   * Returns `true` if the current instance is a superset of the given dataset; differently put: if the given dataset
+   * is a subset of, is contained in the current dataset.
```
Comment (TallTed) [inline src/N3Store.js:776]:
```suggestion
   * Returns `true` if the current dataset is a superset of the given dataset; in other words, returns `true` if
   * the given dataset is a subset of, i.e., is contained within, the current dataset.
```

File: src/N3Store.js [hunk 4]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+      for (const quad of quads)
+        this.add(quad);
+    }
+    return this;
+  }
+
+
+  /**
+   * Returns `true` if the current instance is a superset of the given dataset; differently put: if the given dataset
+   * is a subset of, is contained in the current dataset.
```
Comment (TallTed) [inline src/N3Store.js:793]:
```suggestion
   * Returns `true` if the current dataset is a superset of the given dataset; in other words, returns `true` if
   * the given dataset is a subset of, i.e., is contained within, the current dataset.
```

File: src/N3Store.js [hunk 5]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+   *
+   * Blank Nodes will be normalized.
+   */
+  contains(other) {
+    return other.every(quad => this.has(quad));
+  }
+
+  /**
+   * This method removes the quads in the current instance that match the given arguments.
```
Comment (TallTed) [inline src/N3Store.js:802]:
```suggestion
   * This method removes the quads in the current dataset that match the given arguments.
```

File: src/N3Store.js [hunk 6]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+   *
+   * The logic described in {@link https://rdf.js.org/dataset-spec/#quad-matching|Quad Matching} is applied for each
+   * quad in this dataset to select the quads which will be deleted.
```
Comment (TallTed) [inline src/N3Store.js:805]:
```suggestion
   * quad in this dataset, to select the quads which will be deleted.
```

File: src/N3Store.js [hunk 7]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+   * @param graph     The optional exact graph to match.
+   */
+  deleteMatches(subject, predicate, object, graph) {
+    for (const quad of this.match(subject, predicate, object, graph))
+      this.removeQuad(quad);
+    return this;
+  }
+
+  /**
+   * Returns a new dataset that contains all quads from the current dataset, not included in the given dataset.
```
Comment (TallTed) [inline src/N3Store.js:819]:
```suggestion
   * Returns a new dataset that contains all quads from the current dataset that are not included in the given dataset.
```

File: src/N3Store.js [hunk 8]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+   */
+  difference(other) {
+    return this.filter(quad => !other.has(quad));
+  }
+
+  /**
+   * Returns true if the current instance contains the same graph structure as the given dataset.
```
Comment (TallTed) [inline src/N3Store.js:826]:
```suggestion
   * Returns true if the current dataset contains the same graph structure as the given dataset.
```

File: src/N3Store.js [hunk 9]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+  filter(iteratee) {
+    const store = new N3Store();
+    for (const quad of this)
+      if (iteratee(quad, this))
+        store.add(quad);
+    return store;
+  }
+
+  /**
+   * Returns a new dataset containing alls quads from the current dataset that are also included in the given dataset.
```
Comment (TallTed) [inline src/N3Store.js:848]:
```suggestion
   * Returns a new dataset containing all quads from the current dataset that are also included in the given dataset.
```

File: src/N3Store.js [hunk 10]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+    const store = new N3Store();
+    for (const quad of this)
+      store.add(iteratee(quad, this));
+    return store;
+  }
+
+  /**
+   * This method calls the `iteratee` on each `quad` of the `Dataset`. The first time the `iteratee` is called, the
+   * `accumulator` value is the `initialValue` or, if not given, equals to the first quad of the `Dataset`. The return
+   * value of the `iteratee` is used as `accumulator` value for the next calls.
```
Comment (TallTed) [inline src/N3Store.js:867]:
```suggestion
   * This method calls the `iteratee` method on each `quad` of the `Dataset`. The first time the `iteratee` method
   * is called, the `accumulator` value is the `initialValue`, or, if not given, equals the first quad of the `Dataset`.
   * The return value of each call to the `iteratee` method is used as the `accumulator` value for the next call.
```

File: src/N3Store.js [hunk 11]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+  reduce(callback, initialValue) {
+    const iter = this.readQuads();
+    let accumulator = initialValue === undefined ? iter.next().value : initialValue;
+    for (const quad of iter)
+      accumulator = callback(accumulator, quad, this);
+    return accumulator;
+  }
+
+  /**
+   * Returns the set of quads within the dataset as a host language native sequence, for example an `Array` in
```
Comment (TallTed) [inline src/N3Store.js:882]:
```suggestion
   * Returns the set of quads within the dataset as a host-language-native sequence, for example an `Array` in
```

File: src/N3Store.js [hunk 12]
```diff
@@ -775,6 +771,158 @@ export default class N3Store {
...
+   * ECMAScript-262.
+   *
+   * Since a `Dataset` is an unordered set, the order of the quads within the returned sequence is arbitrary.
+   */
+  toArray() {
+    return this.getQuads();
+  }
+
+  /**
+   * Returns an N-Quads string representation of the dataset, preprocessed with
```
Comment (TallTed) [inline src/N3Store.js:892]:
```suggestion
   * Returns an N-Quads string representation of the dataset, preprocessed with the
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Minor, editorial, human-facing

**Review by RubenVerborgh — APPROVED** [review]
Yes, this is great!

File: src/N3Store.js (hunk 4 above)
Comment (jeswr) [reply src/N3Store.js:793]: @TallTed all phrasing for JSDoc in this PR is taken directly from https://rdf.js.org/dataset-spec/#dataset-interface. You may wish to reflect these changes over there.

**Pull request merged**